2_process/src/calculate_toha_continuous.R

```diff
@@ -0,0 +1,232 @@
+  benthic_area[completely_above_lake] <- 0
+  
+  return(data.frame(Z1 = Z1, Z2 = Z2, habitat = benthic_area))
+}
+
+# Returns vector, not data.frame
+thermal_optical_habitat_area <- function(tha_df, oha_df, hypso) {
+  
+  stopifnot(nrow(tha_df) == nrow(oha_df))
+  
+  # Make a placeholder TOHA vector
+  toha <- rep(NA, nrow(oha_df))
+  
+  ##### Find instances where TOHA is zero #####
+  
+  # Check to see if TOHA is zero because either OHA or THA is zero
+  # Meaning either OHA or THA (or both) are above/below the lake
+  i_THA_OHA_zero <- which(tha_df$habitat == 0 | oha_df$habitat == 0)
+  
+  # Check to see if THA and OHA do not overlap
+  i_THA_above_OHA <- which(tha_df$Z1 <= oha_df$Z1 & tha_df$Z2 <= oha_df$Z1)
+  i_OHA_above_THA <- which(oha_df$Z1 <= tha_df$Z1 & oha_df$Z2 <= tha_df$Z1)
+  
+  # Combine those indices that should have zero TOHA
+  i_zero_toha <- unique(c(i_THA_OHA_zero, i_THA_above_OHA, i_OHA_above_THA))
+  
+  # Add zeros to TOHA vector
+  toha[i_zero_toha] <- 0
+  
+  ##### Find instances where TOHA doesn't need to be calculated #####
+  # This would be times where TOHA = THA or TOHA = OHA
+  
+  # THA inside of OHA, therefore TOHA = THA
+  i_THA_inside_OHA <- which(tha_df$Z1 >= oha_df$Z1 & tha_df$Z2 <= oha_df$Z2)
+  toha[i_THA_inside_OHA] <- tha_df$habitat[i_THA_inside_OHA]
+  
+  # OHA inside of THA, therefore TOHA = OHA
+  i_OHA_inside_THA <- which(oha_df$Z1 >= tha_df$Z1 & oha_df$Z2 <= tha_df$Z2)
+  toha[i_OHA_inside_THA] <- oha_df$habitat[i_OHA_inside_THA]
+  
+  ##### Combine all of the indices where TOHA could be determined without calculating #####
+  i_noncalc <- c(i_zero_toha, i_THA_inside_OHA, i_OHA_inside_THA)
+  to_calc <- which(!seq_len(length(toha)) %in% i_noncalc) # handles instance where i_noncalc is empty
+  
+  if(length(to_calc) > 0) {
+    ##### Calculate benthic area where TOHA couldn't already be determined
+    # The following is only done for those instances where TOHA has yet to 
+    # be determined.
+    
+    # Resuling matrix is 4 x nrow(oha_df[to_calc]):
+    #   Columns: one column per timestep that still needs TOHA calculated
+    #   Row 1: OHA Z1s
+    #   Row 2: OHA Z2s
+    #   Row 3: THA Z1s
+    #   Row 4: THA Z2s
+    
+    all_Zs_to_calc <- t(matrix(c(oha_df$Z1[to_calc], oha_df$Z2[to_calc], 
+                                 tha_df$Z1[to_calc], tha_df$Z2[to_calc]),
+                               ncol = 4))
+    all_Zs_sorted <- apply(all_Zs_to_calc, 2, sort) # sort each column's depths
+    
+    # Find the new Z's - should be the two middle values when you sort each column
+    # So i = 2, and i = 3 (there are 4 columns, Z1 & Z2 for both THA & OHA)
+    Z1 <- as.vector(all_Zs_sorted[2,])
+    Z2 <- as.vector(all_Zs_sorted[3,])
+    
+    A1 <- resample_hypso(hypso, Z1)[["areas"]]
+    A2 <- resample_hypso(hypso, Z2)[["areas"]]
+    
+    ##### Use areas and depths at thresholds to calculate benthic area for TOHA
+    
+    benthic_area <- calc_benthic_area(Z1, Z2, A1, A2)
```

Comment:
I didn't review the other parts of this after noting the assumptions in the `calc_benthic_area` function above - can you address those comments in THA and OHA and then I will take a closer look at OHA?